src/server/keystore/keystore.js

```diff
@@ -0,0 +1,89 @@
+import { writeFileSync, readFileSync, existsSync } from 'fs';
+import { createCipher, createDecipher } from 'crypto';
+import * as errors from './errors';
+
+const VERSION = 1;
+const ALGORITHM = 'aes-256-gcm';
+
+export class Keystore {
+  constructor(path, password = '') {
+    this.path = path;
+    this.data = {};
+    this.password = password;
+
+    this.load();
+  }
+
+  static errors = errors;
+
+  static encrypt(data, password = '') {
+    const cipher = createCipher(ALGORITHM, password);
```

Comment:
I think this should be a bit different and not use `createCipher` with the password directly. It only uses MD5 with a single iteration and no salt. Instead we should use pbkdf2 with a salt and large number of iterations and then create the cipher IV.

From the docs:

> The implementation of crypto.createCipher() derives keys using the OpenSSL function EVP_BytesToKey with the digest algorithm set to MD5, one iteration, and no salt. The lack of salt allows dictionary attacks as the same password always creates the same key. The low iteration count and non-cryptographically secure hash algorithm allow passwords to be tested very rapidly.
>
> In line with OpenSSL's recommendation to use pbkdf2 instead of EVP_BytesToKey it is recommended that developers derive a key and IV on their own using crypto.pbkdf2() and to use crypto.createCipheriv() to create the Cipher object. Users should not use ciphers with counter mode (e.g. CTR, GCM or CCM) in crypto.createCipher(). A warning is emitted when they are used in order to avoid the risk of IV reuse that causes vulnerabilities. For the case when IV is reused in GCM, see Nonce-Disrespecting Adversaries for details.

Reply:
Thanks @jaymode, I have updated the implementation to use createCipheriv with a key derived from pbkdf2.

https://github.com/tylersmalley/kibana/blob/08a74ef4639d0d9ce62d73954b4a39ac26d99c48/src/server/keystore/keystore.js#L20-L52